pkg/statedb2/txn.go

```diff
@@ -0,0 +1,400 @@
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Delete from the primary index first to grab the object.
+	// We assume that "data" has only enough defined fields to
+	// compute the primary key.
+	idKey := meta.primaryIndexer().fromObject(object{data: data}).First()
+	idIndexTree := txn.indexWriteTxn(tableName, meta.primaryIndexer().name)
+	obj, existed := idIndexTree.Delete(idKey)
+	if !existed {
+		return nil, false, fmt.Errorf("object not found")
+	}
+
+	txn.pendingObjectDeltas[tableName]--
+
+	// Update revision index.
+	indexTree := txn.indexWriteTxn(tableName, RevisionIndex)
+	if _, ok := indexTree.Delete(revisionKey(obj.revision, idKey)); !ok {
+		panic("BUG: Revision entry not found")
+	}
+
+	// Then update secondary indexes.
+	for index, indexer := range meta.secondaryIndexers() {
+		indexer.fromObject(obj).Foreach(func(key []byte) {
+			if !indexer.unique {
+				key = append(key, idKey...)
+			}
+			txn.indexWriteTxn(tableName, index).Delete(key)
+		})
+	}
+
+	// And finally insert the object into the graveyard.
+	if txn.hasDeleteTrackers(tableName) {
+		graveyardIndex := txn.indexWriteTxn(tableName, GraveyardIndex)
+		obj.revision = revision
+		if old, existed := graveyardIndex.Insert(idKey, obj); existed {
+			txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Delete(revisionKey(old.revision, idKey))
+			txn.pendingGraveyardDeltas[tableName]--
+		}
+		txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Insert(revisionKey(revision, idKey), obj)
+		txn.pendingGraveyardDeltas[tableName]++
+	}
+
+	return obj.data, true, nil
+}
+
+func (txn *txn) Abort() {
+	txn.db.metrics.WriteTxnDuration.With(prometheus.Labels{
+		"tables":  txn.tableNames,
+		"package": txn.packageName,
+	}).Observe(time.Since(txn.acquiredAt).Seconds())
+
+	if txn.writeTxns == nil {
+		return
+	}
+	txn.smus.Unlock()
+	*txn = zeroTxn
+}
+
+func (txn *txn) Commit() {
+	if txn.writeTxns == nil {
+		return
+	}
+
+	// Commit each individual changed index to each table.
+	// We don't notify yet (CommitOnly) as the root needs to be updated
+	// first.
+	for tableIndex, subTxn := range txn.writeTxns {
+		table, ok := txn.tables[tableIndex.table]
+		if !ok {
+			panic("BUG: Table " + tableIndex.table + " not cached")
+		}
+		table.indexes, _, _ =
+			table.indexes.Insert([]byte(tableIndex.index), subTxn.CommitOnly())
+	}
+
+	db := txn.db
+
+	// Acquire the lock on the root tree to sequence the updates to it.
+	db.mu.Lock()
+	rootTxn := db.root.Load().Txn()
+
+	// Insert the modified tables into the root.
+	for name, table := range txn.tables {
+		rootTxn.Insert([]byte(name), *table)
+	}
+
+	// Commit the new root.
+	newRoot := rootTxn.CommitOnly()
+	db.root.Store(newRoot)
+	db.mu.Unlock()
+
+	// Now that new root is available notify of the changes by closing the watch channels.
+	for _, subTxn := range txn.writeTxns {
+		subTxn.Notify()
+	}
+	rootTxn.Notify()
+
+	txn.smus.Unlock()
+	for name, delta := range txn.pendingObjectDeltas {
```

Comment:
We proccess the pending object deltas here, as we should, but I forgot to add the same logic for the `pendingGraveyardDeltas` :face_exhaling: